irohad/ordering/impl/on_demand_ordering_service_impl.cpp

```diff
@@ -5,24 +5,135 @@
-        transactions) {
-  throw std::logic_error("Function not yet implemented");
+    const CollectionType &transactions) {
+  // read lock
+  std::shared_lock<std::shared_timed_mutex> guard(lock_);
+  log_->info("onTransactions => collections size = {}", transactions.size());
+
+  std::for_each(
+      transactions.begin(), transactions.end(), [this](const auto &obj) {
+        current_proposal_.second.push(obj);
+      });
+  log_->info("onTransactions => collection is inserted");
+}
+
+boost::optional<OnDemandOrderingServiceImpl::ProposalType>
+OnDemandOrderingServiceImpl::onRequestProposal(transport::RoundType round) {
+  // read lock
+  std::shared_lock<std::shared_timed_mutex> guard(lock_);
+  auto proposal = proposal_map_.find(round);
+  if (proposal != proposal_map_.end()) {
+    return proposal->second;
+  } else {
+    return boost::none;
+  }
+}
+
+// ---------------------------------| Private |---------------------------------
+
+void OnDemandOrderingServiceImpl::packNextProposal(
+    RoundOutput outcome, const transport::RoundType &last_round) {
+  log_->info("pack next proposal...size of queue = {}",
+             current_proposal_.second.unsafe_size());
+  if (not current_proposal_.second.empty()) {
+    proposal_map_.insert(
+        std::make_pair(current_proposal_.first, emitProposal()));
+    log_->info("packNextProposal: data has been fetched");
+  }
+
+  round_queue_.push(current_proposal_.first);
+  log_->info("packNextProposal: pushed in queue");
+
+  auto current_round = current_proposal_.first;
+  decltype(current_round) next_round;
+  switch (outcome) {
+    case RoundOutput::SUCCESSFUL:
+      next_round = std::make_pair(current_round.first + 1, 1);
+      break;
+    case RoundOutput::REJECT:
+      next_round =
+          std::make_pair(current_round.first, current_round.second + 1);
+      break;
+  }
+
+  log_->info("nextRound is: [{}, {}]", next_round.first, next_round.second);
+
+  current_proposal_.first = next_round;
+  current_proposal_.second.clear();  // operator = of tbb::queue is closed
 }
 
-boost::optional<std::shared_ptr<shared_model::interface::Proposal>>
-OnDemandOrderingServiceImpl::onRequestProposal(uint64_t round) {
-  throw std::logic_error("Function not yet implemented");
+OnDemandOrderingServiceImpl::ProposalType
+OnDemandOrderingServiceImpl::emitProposal() {
+  auto mutable_proposal = shared_model::proto::ProposalBuilder()
+                              .height(current_proposal_.first.first)
+                              .createdTime(iroha::time::now());
+  log_->info("Mutable proposal created");
+
+  TransactionType current_tx;
+  using ProtoTxType = shared_model::proto::Transaction;
+  std::vector<TransactionType> collection;
+
+  // outer method should guarantee availability of at least one transaction in
+  // queue
+  while (current_proposal_.second.try_pop(current_tx)
+         and collection.size() < transaction_limit_) {
+    collection.emplace_back(current_tx);
```

Comment:
`onTransactions` does not enforce the size of the queue, so `current_proposal_` can contain more transactions than `transaction_limit`. Please comment if this is wrong, or the case should be handled to avoid missing transactions.